[WebAssembly] Fix tests and build on x86-64 Linux with BBQJIT

takikawa opened the pull request:
#### 361c5fba483c9c157cf689d4e60acaccc45d74c5
<pre>
[WebAssembly] Fix tests and build on x86-64 Linux with BBQJIT
<a href="https://bugs.webkit.org/show_bug.cgi?id=253618">https://bugs.webkit.org/show_bug.cgi?id=253618</a>

Reviewed by Yusuke Suzuki.

This patch first enables building of the BBQJIT on x86-64 Linux
platforms by adding a line in WebKitFeatures.cmake.

The rest of the patch fixes build issues with GCC.

  - GCC doesn&apos;t support std::fabsf so use std::fabs instead.
    (<a href="https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79700)">https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79700)</a>
  - GCC doesn&apos;t support explicit specialization in class scope so
    this patch uses a `if constexpr` workaround.
    (<a href="https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85282)">https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85282)</a>
  - Add RELEASE_ASSERT_NOT_REACHED to suppress warnings.

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::Location::fromArgumentLocation):
(JSC::Wasm::BBQJIT::loadOpForTypeKind):
(JSC::Wasm::BBQJIT::storeOpForTypeKind):
(JSC::Wasm::BBQJIT::addF32Copysign):
(JSC::Wasm::BBQJIT::fromJSCReg):
(JSC::Wasm::BBQJIT::fromJSCReg&lt;RegisterID&gt;): Deleted.
(JSC::Wasm::BBQJIT::fromJSCReg&lt;FPRegisterID&gt;): Deleted.
* Source/cmake/WebKitFeatures.cmake:

Canonical link: <a href="https://commits.webkit.org/261417@main">https://commits.webkit.org/261417@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d54e2997f09ab55bdd02eebe0529a938bc59bc5b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111664 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20800 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/276 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3418 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/120401 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22155 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11880 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/3157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117430 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99599 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/104572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45319 "layout-tests (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/100163 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13279 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/175 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/87754 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/11403 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13769 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/101385 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19214 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52173 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31530 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7948 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15759 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/109424 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26934 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -1684,6 +1688,8 @@ class BBQJIT {
         case TypeKind::Void:
             RELEASE_ASSERT_NOT_REACHED();
             return StoreOpType::I64Store;
+        default:
+            RELEASE_ASSERT_NOT_REACHED();
```
Can you place it outside of switch? We would like to list up all types here, and `default` suppress this compiler check, so not use `default`.

takikawa replied:
Thanks, fixed.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -1609,6 +1611,8 @@ class BBQJIT {
             RELEASE_ASSERT_NOT_REACHED();
             return LoadOpType::I64Load;
+        default:
+            RELEASE_ASSERT_NOT_REACHED();
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -162,6 +162,8 @@ class BBQJIT {
             case ValueLocation::Kind::Stack:
                 return Location::fromStack(argLocation.location.offsetFromFP());
+            default:
+                RELEASE_ASSERT_NOT_REACHED();
```
Ditto.

Constellation reviewed (approved):
r=me with comments.

takikawa committed:
[WebAssembly] Fix tests and build on x86-64 Linux with BBQJIT
https://bugs.webkit.org/show_bug.cgi?id=253618

Reviewed by Yusuke Suzuki.

This patch first enables building of the BBQJIT on x86-64 Linux
platforms by adding a line in WebKitFeatures.cmake.

The rest of the patch fixes build issues with GCC.

  - GCC doesn't support std::fabsf so use std::fabs instead.
    (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79700)
  - GCC doesn't support explicit specialization in class scope so
    this patch uses a `if constexpr` workaround.
    (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85282)
  - Add RELEASE_ASSERT_NOT_REACHED to suppress warnings.

* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::Location::fromArgumentLocation):
(JSC::Wasm::BBQJIT::loadOpForTypeKind):
(JSC::Wasm::BBQJIT::storeOpForTypeKind):
(JSC::Wasm::BBQJIT::addF32Copysign):
(JSC::Wasm::BBQJIT::fromJSCReg):
(JSC::Wasm::BBQJIT::fromJSCReg<RegisterID>): Deleted.
(JSC::Wasm::BBQJIT::fromJSCReg<FPRegisterID>): Deleted.
* Source/cmake/WebKitFeatures.cmake:

Canonical link: https://commits.webkit.org/261417@main
